File: apina-core/src/test/java/fi/evident/apina/java/model/JavaModelTest.java
package fi.evident.apina.java.model;

import fi.evident.apina.java.model.type.JavaBasicType;
import fi.evident.apina.java.reader.ReflectionClassMetadataLoader;
import org.junit.Test;

import java.lang.annotation.Retention;
import java.util.Set;

import static java.lang.annotation.RetentionPolicy.RUNTIME;
import static org.hamcrest.CoreMatchers.hasItem;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class JavaModelTest {

    @Test
    public void impliedAnnotations() {
        JavaModel model = ReflectionClassMetadataLoader.loadClassesFromInheritanceTree(MyClass.class);

        JavaBasicType metaMeta = new JavaBasicType(MyMetaMetaAnnotation.class);

        Set<JavaBasicType> annotations = model.findAnnotationsImpliedBy(metaMeta);
        assertThat(annotations.size(), is(3));
        assertThat(annotations, hasItem(new JavaBasicType(MyMetaMetaAnnotation.class)));
        assertThat(annotations, hasItem(new JavaBasicType(MyMetaAnnotation.class)));
        assertThat(annotations, hasItem(new JavaBasicType(MyAnnotation.class)));
    }

    @Retention(RUNTIME)
    private @interface MyMetaMetaAnnotation { }

    @Retention(RUNTIME)
    @MyMetaMetaAnnotation
    private @interface MyMetaAnnotation { }

    @Retention(RUNTIME)
    @MyMetaAnnotation
    private @interface MyAnnotation { }

    @MyAnnotation
    private interface InterfaceWithAnnotation { }

    @MyAnnotation
    private static class MyClass implements InterfaceWithAnnotation { }
}

File: apina-core/src/test/kotlin/fi/evident/apina/java/model/JavaModelTest.kt
package fi.evident.apina.java.model

import fi.evident.apina.java.model.type.JavaType
import fi.evident.apina.java.reader.ReflectionClassMetadataLoader.loadClassesFromInheritanceTree
import org.hamcrest.CoreMatchers.hasItem
import org.junit.Assert.assertThat
import org.junit.Test
import kotlin.annotation.AnnotationRetention.RUNTIME
import kotlin.test.assertEquals

class JavaModelTest {

    @Test
    fun impliedAnnotations() {
        val model = loadClassesFromInheritanceTree(MyClass::class.java)

        val metaMeta = JavaType.Basic(MyMetaMetaAnnotation::class.java)

        val annotations = model.findAnnotationsImpliedBy(metaMeta)
        assertEquals(3, annotations.size)
        assertThat(annotations, hasItem(JavaType.Basic(MyMetaMetaAnnotation::class.java)))
        assertThat(annotations, hasItem(JavaType.Basic(MyMetaAnnotation::class.java)))
        assertThat(annotations, hasItem(JavaType.Basic(MyAnnotation::class.java)))
    }

    @Retention(RUNTIME)
    private annotation class MyMetaMetaAnnotation

    @Retention(RUNTIME)
    @MyMetaMetaAnnotation
    private annotation class MyMetaAnnotation

    @Retention(RUNTIME)
    @MyMetaAnnotation
    private annotation class MyAnnotation

    @MyAnnotation
    private interface InterfaceWithAnnotation

    @MyAnnotation
    private class MyClass : InterfaceWithAnnotation
}
